[MNOE-688] Generate getter and setter for has_one relationship

x4d3 opened the pull request:
This PR introduces 2 big changes.

- The use of a has_one relationship helper
- Stop loading everything in current_user

## has_one relationship helper
This PR introduces an helper to easily create elements with relationships. 

```
o = OrgaRelation.new(role: 'Super Admin')
o.relationships.user = User.new(id: user_id)
o.relationships.organization = User.new(id: organization_id)
o.save!
```
is replaced by
 ```
OrgaRelation.create!(user: user, organization: organization, role: 'Super Admin')
```

## make current_user call smaller

In the current version, current_user is preloaded with a lot of dependencies, which are not always used. This can introduce problems of performances, caching and pagination. The objective of this PR is too stop relying on the autoload of theses dependencies and let the controllers manage the loading of what they need.

x4d3 committed:
[MNOE-688] Generate relation_id getter and setter for has_one relationship

- Introduce HasOneExtension, generating getter and setter for has_one relations, making them saved in relationships, and allowing their retrieving from a relations cache
- Remove useless timestamp method
- Rplace is_integer by is_id that is more semantic
- Try to not rely on parent_organization

ouranos commented on core/lib/json_api_client_extension/has_one_extension.rb:
```diff
@@ -0,0 +1,27 @@
+  class_methods do
+    def has_one(attr_name, options = {})
+      class_eval <<-CODE
+        def #{attr_name}_id=(id)
```
Coud you add something like:
```ruby
ActiveSupport::Deprecation.warn('Use relationships instead')
# => "DEPRECATION WARNING: Use relationships instead (called from your_code.rb:1)"
```

This will allow us to spot and progressively replace all the calls

alachaum replied:
👍

x4d3 committed:
Improve performance by stopping loading everything in current_user

Each controller must be in charge of managing its own payload.

x4d3 committed:
Fix spec

manu-d commented on api/app/controllers/mno_enterprise/jpi/v1/admin/impac/dashboard_templates_controller.rb:
```diff
@@ -19,61 +22,44 @@ def index
-    def dashboard_template
-      @dashboard_template ||= dashboard_templates.find(params[:id].to_i).first
+    def load_organizations
+      @organizations = MnoEnterprise::Organization.where('user.ids': current_user.id)
```
```ruby
@organizations = MnoEnterprise::Organization.where('users.id': current_user.id)
```

manu-d commented on api/lib/mno_enterprise/concerns/controllers/jpi/v1/admin/base_resource_controller.rb:
```diff
@@ -14,12 +14,8 @@ module MnoEnterprise::Concerns::Controllers::Jpi::V1::Admin::BaseResourceControl
-
   def parent_organization
-    @parent_organization ||= current_user.organizations.to_a.find { |o| o.id.to_s == params[:organization_id].to_s }
+    @parent_organization ||= MnoEnterprise::Organization.where(user_id: current_user.id, id: params[:organization_id]).first
```
```ruby
@parent_organization ||= MnoEnterprise::Organization.where('users.id': current_user.id, id: params[:organization_id]).first
```

manu-d commented on api/lib/mno_enterprise/concerns/controllers/jpi/v1/admin/base_resource_controller.rb:
```diff
@@ -33,14 +29,14 @@ def check_authorization
 
   def render_forbidden_request(attempted_action)
-    render json: { errors: {message: "Error while trying to #{attempted_action}: you do not have permission", code: 403} }, status: :forbidden
+    render json: { errors: { message: "Error while trying to #{attempted_action}: you do not have permission", code: 403 } }, status: :forbidden
```
Can't you put this 3 methods in application_controller.rb, so that you dont' have to repeat them as well in the non_admin base controller?

x4d3 replied:
Done.

manu-d commented on api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/dashboard_templates_controller.rb:
```diff
@@ -18,5 +17,6 @@ module MnoEnterprise::Concerns::Controllers::Jpi::V1::Impac::DashboardTemplatesC
   # GET /mnoe/jpi/v1/impac/dashboard_templates
   def index
     @templates = MnoEnterprise::Dashboard.published_templates.includes(*DASHBOARD_DEPENDENCIES)
+    @organizations = MnoEnterprise::Organization.where('user.ids': current_user.id)
```
```ruby
@organizations = MnoEnterprise::Organization.where('users.id': current_user.id)
```

manu-d commented on api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/dashboards_controller.rb:
```diff
@@ -18,11 +18,13 @@ module MnoEnterprise::Concerns::Controllers::Jpi::V1::Impac::DashboardsControlle
   # GET /mnoe/jpi/v1/impac/dashboards
   def index
     dashboards
+    @organizations = MnoEnterprise::Organization.where('user.ids': current_user.id)
```
```ruby
@organizations = MnoEnterprise::Organization.where('users.id': current_user.id)
```

manu-d replied:
```ruby
@organizations = MnoEnterprise::Organization.where('users.id': current_user.id)
```

manu-d replied:
```ruby
@organizations = MnoEnterprise::Organization.where('users.id': current_user.id)
```

manu-d commented on api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/dashboards_controller.rb:
```diff
@@ -18,11 +18,13 @@ module MnoEnterprise::Concerns::Controllers::Jpi::V1::Impac::DashboardsControlle
   # GET /mnoe/jpi/v1/impac/dashboards/1
   #   -> GET /api/mnoe/v1/users/1/dashboards
   def show
+    @organizations = MnoEnterprise::Organization.where('user.ids': current_user.id)
```
```ruby
@organizations = MnoEnterprise::Organization.where('users.id': current_user.id)
```

manu-d commented on api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/dashboards_controller.rb:
```diff
@@ -35,6 +37,7 @@ def create
     @dashboard = MnoEnterprise::Dashboard.create!(dashboard_create_params)
     MnoEnterprise::EventLogger.info('dashboard_create', current_user.id, 'Dashboard Creation', @dashboard)
     @dashboard = dashboard.load_required(*DASHBOARD_DEPENDENCIES)
+    @organizations = MnoEnterprise::Organization.where('user.ids': current_user.id)
```
```ruby
@organizations = MnoEnterprise::Organization.where('users.id': current_user.id)
```

manu-d commented on api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/dashboards_controller.rb:
```diff
@@ -46,7 +49,7 @@ def update
     # authorize! :manage_dashboard, dashboard
     dashboard.update_attributes!(dashboard_update_params)
-
+    @organizations = MnoEnterprise::Organization.where('user.ids': current_user.id)
```
```ruby
@organizations = MnoEnterprise::Organization.where('users.id': current_user.id)
```

manu-d replied:
```ruby
@organizations = MnoEnterprise::Organization.where('users.id': current_user.id)
```

manu-d commented on api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/dashboards_controller.rb:
```diff
@@ -73,6 +76,7 @@ def copy
     # Owner is the current user by default, can be overriden to something else (eg: current organization)
     @dashboard = template.copy!(current_user, dashboard_params[:name], dashboard_params[:organization_ids])
     @dashboard = @dashboard.load_required(DASHBOARD_DEPENDENCIES)
+    @organizations = MnoEnterprise::Organization.where('user.ids': current_user.id)
```
```ruby
@organizations = MnoEnterprise::Organization.where('users.id': current_user.id)
```

manu-d commented on api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/widgets_controller.rb:
```diff
@@ -17,7 +17,7 @@ module MnoEnterprise::Concerns::Controllers::Jpi::V1::Impac::WidgetsController
   def index
     render_not_found('organization') unless parent_organization
-    @widgets = MnoEnterprise::Widget.find(organization_id: parent_organization.id)
+    @widgets = MnoEnterprise::Widget.find(organization_id: parent_organization_id)
```
parent_organization_id can be an uid as well
```ruby
@widgets = MnoEnterprise::Widget.where('organization.id': parent_organization.id)
```

manu-d commented on api/lib/mno_enterprise/concerns/controllers/jpi/v1/subscriptions_controller.rb:
```diff
@@ -34,15 +34,15 @@ def create
+    authorize! :manage_app_instances, orga_relation
 
-    subscription = MnoEnterprise::Subscription.where(organization_id: parent_organization.id, id: params[:id]).first
+    subscription = MnoEnterprise::Subscription.where(organization_id: parent_organization_id, id: params[:id]).first
```
```ruby
subscription = MnoEnterprise::Subscription.where('organization.id': parent_organization_id, id: params[:id]).first
```

manu-d commented on api/lib/mno_enterprise/concerns/controllers/jpi/v1/subscriptions_controller.rb:
```diff
@@ -55,20 +55,20 @@ def update
+    authorize! :manage_app_instances, orga_relation
 
-    subscription = MnoEnterprise::Subscription.where(organization_id: parent_organization.id, id: params[:id]).first
+    subscription = MnoEnterprise::Subscription.where(organization_id: parent_organization_id, id: params[:id]).first
```
```ruby
subscription = MnoEnterprise::Subscription.where('organization.id': parent_organization_id, id: params[:id]).first
```

x4d3 committed:
Fix 'user.ids' -> 'users.id'

x4d3 committed:
Add only needed fields for users

x4d3 committed:
Moved common method to application_controller

x4d3 committed:
Moved common method back to base_resource

ouranos commented on api/app/controllers/mno_enterprise/jpi/v1/admin/impac/dashboard_templates_controller.rb:
```diff
@@ -1,6 +1,8 @@
 module MnoEnterprise
   class Jpi::V1::Admin::Impac::DashboardTemplatesController < Jpi::V1::Admin::BaseResourceController
 
+    before_action :load_organizations, except: [:destroy]
```
Where's `@organizations` used? Only in the view?
With the `before_action` you're loading organizations all the time, even if the `find_one!` or `save!` is going to fail

x4d3 replied:
True. I'll update it.

ouranos commented on api/lib/mno_enterprise/concerns/controllers/jpi/v1/marketplace_controller.rb:
```diff
@@ -59,13 +59,13 @@ def app_relation(org_id = nil)
   end
+
+  def orga_relation
+    MnoEnterprise::OrgaRelation.where('user.id' => current_user.id, 'organization.id': params[:organization_id]).first
```
Do you need the full orga_relation here? Otherwise just add `select(:id)`?

ouranos commented on core/lib/json_api_client_extension/has_one_extension.rb:
```diff
@@ -0,0 +1,29 @@
+      class_eval <<-CODE
+        def #{attr_name}_id=(id)
+          # Uncomment when doing refactoring
+          # ActiveSupport::Deprecation.warn(self.class.name + ".#{attr_name}_id= Use relationships instead")
```
This should  not be commented?

x4d3 replied:
it should be not commented during development, but not for prod. I will add a check.

ouranos commented on core/lib/json_api_client_extension/json_api_client_orm_adapter.rb:
```diff
@@ -15,14 +15,14 @@ def column_names
     # @see OrmAdapter::Base#get!
     def get!(id)
-      res = klass.includes(*klass::INCLUDED_DEPENDENCIES).find(wrap_key(id)).first
+      res = klass.find(wrap_key(id)).first
```
Are you removing the `INCLUDED_DEPENDENCIES` feature? Was it widely used or just for the current_user?

x4d3 replied:
it was mainly used for the current_user, but also for other controllers. (for example when `current_user.organizations` was called)

ouranos commented on core/lib/mno_enterprise/concerns/models/organization.rb:
```diff
@@ -94,11 +94,6 @@ def add_user!(user, role = 'Member')
     MnoEnterprise::OrgaRelation.create!(organization_id: self.id, user_id: user.id, role: role)
   end
 
-  def provision_app_instance!(app_nid)
```
see  #584 
This is going to introduce regressions

ouranos reviewed (commented):
See comments.

Have you done some local performance testing?
I have the feeling that, in some cases, it might not be faster as you're trading a smaller response for more API requests.

x4d3 commented:
I will check the performance. The problem is that includes does not really work in v2, as it does not paginates. On top of that the MnoHub calls can be very long to build up the user with all the dependencies. It's definitely more interesting to make more smaller calls, easy to cache and reproduce than complicated calls with multiple includes.

x4d3 committed:
Do not load_organizations in before_action

x4d3 committed:
Apply code review comments

- fix issue with provisioning + update spec
- Add log warning in has_one in test and development

ouranos commented on api/lib/mno_enterprise/concerns/controllers/jpi/v1/deletion_requests_controller.rb:
```diff
@@ -59,4 +58,5 @@ def destroy
       head :bad_request
     end
   end
+
```
Empty line :policeman:

ouranos reviewed (commented):
LGTM

`core/lib/mno_enterprise/concerns/controllers/auth/omniauth_callbacks_controller.rb` still calls `provision_app_instance!`

(Might need an extra spec if this hasn't been picked up)

x4d3 committed:
Fix and spec OmniauthCallbacksController.setup_apps

alachaum reviewed (dismissed):
👍  looks good to me
